Gateway/Response/VaultDetailsHandler.php

```diff
@@ -48,9 +48,148 @@ public function __construct(Vault $vaultHelper)
+     * @return PaymentTokenInterface|null
+     */
+    private function getVaultPaymentToken(array $response, $payment)
     {
         if (empty($response['additionalData'])) {
-            return;
+            return null;
+        }
+
+        $additionalData = $response['additionalData'];
+
+        $paymentToken = null;
+
+        foreach (self::ADDITIONAL_DATA_ERRORS as $key => $errorMsg) {
+            if (empty($additionalData[$key])) {
+                $this->adyenLogger->error($errorMsg);
+                return null;
+            }
+        }
+
+        try {
+            // Check if paymentToken exists already
+            $paymentToken = $this->paymentTokenManagement->getByGatewayToken(
+                $additionalData[self::RECURRING_DETAIL_REFERENCE],
+                $payment->getMethodInstance()->getCode(),
+                $payment->getOrder()->getCustomerId()
+            );
+
+            $paymentTokenSaveRequired = false;
+
+            // In case the payment token does not exist, create it based on the additionalData
+            if ($paymentToken === null) {
+                /** @var PaymentTokenInterface $paymentToken */
+                $paymentToken = $this->paymentTokenFactory->create(
+                    PaymentTokenFactoryInterface::TOKEN_TYPE_CREDIT_CARD
+                );
+
+                $paymentToken->setGatewayToken($additionalData[self::RECURRING_DETAIL_REFERENCE]);
+
+                if (strpos($additionalData[self::PAYMENT_METHOD], "paywithgoogle") !== false
+                    && !empty($additionalData['paymentMethodVariant'])) {
+                    $additionalData[self::PAYMENT_METHOD] = $additionalData['paymentMethodVariant'];
+                    $paymentToken->setIsVisible(false);
+                }
+            } else {
+                $paymentTokenSaveRequired = true;
+            }
+
+            $paymentToken->setExpiresAt($this->getExpirationDate($additionalData[self::EXPIRY_DATE]));
+
+            $details = [
+                'type' => $additionalData[self::PAYMENT_METHOD]
+            ];
+
+            if (!empty($additionalData[self::CARD_SUMMARY])) {
+                $details['maskedCC'] =  $additionalData[self::CARD_SUMMARY];
+            }
+
+            if (!empty($additionalData[self::EXPIRY_DATE])) {
+                $details['expirationDate'] =  $additionalData[self::EXPIRY_DATE];
+            }
+
+            $paymentToken->setTokenDetails(json_encode($details));
+
+            // If the token is updated, it needs to be saved to keep the changes
+            if ($paymentTokenSaveRequired) {
+                $this->paymentTokenRepository->save($paymentToken);
+            }
+        } catch (\Exception $e) {
+            $this->adyenLogger->error(print_r($e, true));
+        }
+
+        return $paymentToken;
+    }
+
+    /**
+     * @param $expirationDate
+     * @return string
+     * @throws \Exception
+     */
+    private function getExpirationDate($expirationDate)
```

Comment:
Duplicated function here and in Cron.php. Will it be outsourced with the 3DS vault PR? @msilvagarcia

Reply:
Nevermind, I see it here #814